internal/command/deploy/machinebasedtest.go

```diff
@@ -0,0 +1,240 @@
+package deploy
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/samber/lo"
+	fly "github.com/superfly/fly-go"
+	"github.com/superfly/flyctl/helpers"
+	"github.com/superfly/flyctl/internal/appconfig"
+	"github.com/superfly/flyctl/internal/machine"
+	"github.com/superfly/flyctl/internal/statuslogger"
+	"github.com/superfly/flyctl/internal/tracing"
+	"github.com/superfly/flyctl/iostreams"
+)
+
+type createdTestMachine struct {
+	mach *fly.Machine
+	err  error
+}
+
+func (md *machineDeployment) runTestMachines(ctx context.Context, machineToTest *fly.Machine) (err error) {
+	ctx, span := tracing.GetTracer().Start(ctx, "run_test_machine")
+	var (
+		flaps = md.flapsClient
+		io    = md.io
+	)
+	defer func() {
+		if err != nil {
+			tracing.RecordError(span, err, "failed to run test machine")
+		}
+		span.End()
+	}()
+
+	machineChecks := lo.FlatMap(md.appConfig.AllServices(), func(svc appconfig.Service, _ int) []*appconfig.ServiceMachineCheck {
+		return svc.MachineChecks
+	})
+
+	if len(machineChecks) == 0 {
+		span.AddEvent("no machine checks")
+		return nil
+	}
+
+	machines := lo.Map(machineChecks, func(machineCheck *appconfig.ServiceMachineCheck, _ int) createdTestMachine {
+		image := lo.Ternary(machineCheck.Image == "", md.img, machineCheck.Image)
+		defer func() {
+			if err != nil {
+				statuslogger.Failed(ctx, err)
+			}
+		}()
+
+		mach, err := md.createTestMachine(ctx, machineCheck.Command, image, machineCheck.Entrypoint, machineToTest)
+		if err != nil {
+			err = fmt.Errorf("error running test machine %s: %w", machineCheck.Command, err)
+		}
+
+		return createdTestMachine{mach, err}
+	})
+
+	if m, hasErr := lo.Find(machines, func(m createdTestMachine) bool {
+		return m.err != nil
+	}); hasErr {
+		err := fmt.Errorf("error creating test machine %s: %w", m.mach.ID, m.err)
+		tracing.RecordError(span, err, "failed to create test machine")
+		return err
+	}
+
+	machineSet := machine.NewMachineSet(flaps, io, lo.FilterMap(machines, func(m createdTestMachine, _ int) (*fly.Machine, bool) {
+		if m.err != nil {
+			tracing.RecordError(span, m.err, "failed to create test machine")
+			statuslogger.LogStatus(ctx, statuslogger.StatusFailure, fmt.Sprintf("failed to create test machine: %s", m.err))
+		}
+		return m.mach, m.err == nil
+	}))
+
+	// FIXME: consolidate this wait stuff with deploy waits? Especially once we improve the output
+	err = md.waitForTestMachinesToFinish(ctx, machineSet)
+	if err != nil {
+		tracing.RecordError(span, err, "failed to wait for test cmd machine")
+		return err
+	}
+
+	time.Sleep(2 * time.Second) // Wait 2 secs to be sure logs have reached OpenSearch
```

Comment:
Is there a better way to check for this rather than sleeping?

Reply:
we're now waiting for logs using backoff :)